DEVELOPERS.md

```diff
@@ -40,6 +40,57 @@ Testing setup:
 - `export PARQUET_TEST_DATA=$(pwd)/parquet-testing/data/`
 - `export ARROW_TEST_DATA=$(pwd)/testing/data/`
 
+
+## Test Organization
+
+DataFusion has several levels of tests in its [Test
+Pyramid](https://martinfowler.com/articles/practical-test-pyramid.html)
+and tries to follow [Testing Organization](https://doc.rust-lang.org/book/ch11-03-test-organization.html) in the The Book.
+
+This section highlights the most important test modules that exist
+
+### Unit tests
+
+Tests for the code in an individual module are defined in the same source file with a `test` module, following Rust convention
+
+For example, the logic for pruning predicates is defined in [pruning.rs](https://github.com/apache/arrow-datafusion/blob/main/datafusion/src/physical_optimizer/pruning.rs#L708) using the
+
+```rust
+#[cfg(test)]
+mod tests {
+...
+}
+```
+
+
+### Rust Integration Tests
+
+There are several tests of the public interface of the DataFusion library in the [tests](https://github.com/apache/arrow-datafusion/blob/master/datafusion/tests) directory.
+
+You can run these tests individually using a command such as
+
+```shell
+cargo test -p datafusion --tests sql
+```
+
+The tests are:
```

Comment:
One thing I was thinking @hntd187  and @matthewmturner  would be to put some guidance here like "tests that run a full sql query likely belong in `sql.rs`" (also perhaps describe which tests should go in the "SQL / Postgres integration tests" and which are good to be in sql.rs)

Reply:
Agree.

Reply:
I was going to grain it down a bit more with a module for sql but then specific modules separated by functionality such as "order_by", "group_by" things of that nature, where generally they fall under sql, but more specifically if they touch specific functionality they go deeper. I'm not sure what @matthewmturner has started so maybe I'll wait a day to see what he says.

I also wanted to separate tests for internal things but the unit tests inside individual code seem fine for that.

Reply:
@alamb @hntd187 sounds good.  i have not had a chance to start on this yet and likely wont be able to spend much time on it until i finish some other PRs.  @hntd187 it seems like you have a good plan and id be happy to follow your lead.  assuming that breaking out the sql functionality like that is the agreed upon approach maybe we could make a parent issue with sub tasks for each functionality /  internal thing and  then i could pick up on that once i have some more time?

Reply:
Sure I've started moving things around, and I guess I'm wondering how deep down the rabbit hole do we wanna go here? There are a good number of tests in `sql.rs` that by their name cross considerations, like is reading parquet as a format something for sql? In addition a lot of tests are prepended with csv, which seems somewhat redundant to me, is there a purpose behind it? I think I will reorganize the tests themselves first, touch no code other than moving around and then we can talk about renaming or any actual code changes.

Reply:
> In addition a lot of tests are prepended with csv, which seems somewhat redundant to me, is there a purpose behind it? 

Not that I can remember

> I think I will reorganize the tests themselves first, touch no code other than moving around and then we can talk about renaming or any actual code changes.

I think that is an excellent strategy 👍